core/src/main/java/org/infinispan/reactive/publisher/impl/ClusterPublisherManagerImpl.java

```diff
@@ -217,6 +217,28 @@ public void stop() {
       localStage.whenComplete(biConsumer);
    }
 
+   private <I, R> void handleNoTargets(Function<? super Publisher<I>, ? extends CompletionStage<R>> transformer,
+         Function<? super Publisher<R>, ? extends CompletionStage<R>> finalizer,
+         FlowableProcessor<R> flowableProcessor) {
+      // Need to do this in case if the transformer or finalizer has additional values such as reduce with identity
+      // or switchIfEmpty etc.
+      CompletionStage<R> transformedStage = transformer.apply(Flowable.empty());
+      if (finalizer != null) {
+         transformedStage = transformedStage.thenApply(value -> finalizer.apply(Flowable.just(value)))
+               .thenCompose(Function.identity());
+      }
```

Comment:
I'm not really sure why the application of `finalizer` is needed here. My original PR war aligned to the implementation of `handleContextInvocation` and there finalizer is also not used.

Maybe the finalizer anyway applies in the end since this line is in place in `ClusterPublisherManagerImpl.reduction()`?
```
      // Apply the finalizer first (which subscribes) before emitting items, to avoid buffering in UnicastProcessor
      CompletionStage<R> stage = finalizer.apply(flowableProcessor);
```

Reply:
Finalizer is designed to be allowed to run as many times as you have required subset of entries. In this case, it is extraneous for the call as you pointed out.

To describe the use of finalizer more, it is used in the place you mentioned to consolidate responses from each of the nodes into a single value. However, the finalizer must also be used on each of the remote/local nodes if the reduction is parallel or we have more strict GUARANTEEs which are done per segment. But since we have a single value, it isn't really needed at all tbh :)

Reply:
Thanks for the explanation! Reactive code really drives me crazy sometimes ;-)